src/parser.ts

```diff
@@ -662,6 +672,153 @@ const TRAVERSE_BUILDER: MarkVisitor<(rhs: TraversalResult | null) => TraversalRe
   },
 }
 
+const SELECTOR_BUILDER: MarkVisitor<null> = {
+  group(p) {
+    p.process(SELECTOR_BUILDER)
+    return null
+  },
+
+  everything() {
+    throw new Error('Invalid selector syntax')
+  },
+
+  this() {
+    throw new Error('Invalid selector syntax')
+  },
+
+  parent() {
+    throw new Error('Invalid selector syntax')
+  },
+
+  dblparent(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  traverse(p) {
+    p.process(SELECTOR_BUILDER)
+    while (p.getMark().name !== 'traversal_end') {
+      p.process(TRAVERSE_BUILDER)
+    }
+
+    p.shift()
+    return null
+  },
+
+  this_attr(p) {
+    p.processString()
+    return null
+  },
+
+  neg(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  pos(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  add(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  sub(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  mul(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  div(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  mod(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  pow(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  comp(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  in_range(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  str(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  integer(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  float(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  sci(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  object(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  array(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  tuple(p) {
+    // This should only throw an error until we add support for tuples in selectors.
+    throw new Error('Invalid selector syntax')
+  },
+
+  func_call(p) {
+    p.unshift()
+    const func = p.process(EXPR_BUILDER) as NodeTypes.FuncCallNode
```

Comment:
Aha! Then we can be more explicit here:

```suggestion
  func_call(p, mark) {
    const func = EXPR_BUILDER.func_call(p, mark) as NodeTypes.FuncCallNode
```

Reply:
Ahhh clever!